resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/DriverConfigPropertiesStep.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.k8s.submit.steps
+
+import java.io.StringWriter
+import java.util.Properties
+
+import io.fabric8.kubernetes.api.model._
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.k8s.Config._
+import org.apache.spark.deploy.k8s.Constants._
+import org.apache.spark.deploy.k8s.submit.KubernetesDriverSpec
+
+/**
+ * Create a config map with the driver configuration and attach it to the pod. This needs to
+ * come at the end of the driver configuration so that all modifications to the Spark config
+ * are reflected in the generated config map.
+ */
+private[spark] class DriverConfigPropertiesStep(resourceNamePrefix: String)
+    extends DriverConfigurationStep {
+
+  override def configureDriver(spec: KubernetesDriverSpec): KubernetesDriverSpec = {
+    val configMapName = s"$resourceNamePrefix-driver-conf-map"
+    val configMap = buildConfigMap(configMapName, spec.driverSparkConf)
```

Comment:
it's risky to pass the driver spark configuration this way. The way this is built, this step must be specifically placed last in the list of steps returned by the orchestrator. Ideally, the orchestrators should be able to return monads that can be applied in any order. But this is done in such a way that the driver spark configuration must be completely built in the spec before this step is run.

I think this code shouldn't necessarily even live in a config step, but instead are run directly in the outer submission client. This is what we were doing before, when we just passed all of the driver Spark configurations as system properties to the main class. We can replace that instead with passing these values as `--conf` options. Thoughts?

Reply:
One other question - how are we passing `spark.driver.extraJavaOptions` properties to the driver now?

Reply:
I agree in that it should be in any order. I will restructure it as such.

Reply:
I will also include `--conf spark.driver.extraJavaOptions=` in the spark-submit launched in the driver

Reply:
I see - using the config map is a good idea then, but it still can't be in a config step because it has to assume that all the properties its given are the absolute final ones. We should keep the driver properties code in the same _place_ as it was before, which was separate from the config steps and after all the steps were processed. The implementation can still be with the config map which allows spark-submit to pick up these properties.

Just to double check, will this logic need to be re-used anywhere? I don't think the executor pods have to mount the same properties config map.

> I will also include --conf spark.driver.extraJavaOptions= in the spark-submit launched in the driver

That won't work because by the time spark-submit is running, you'll want to _already_ have these JVM options applied since spark-submit will be running the driver JVM itself. You'll need to find a way to get the spark-submit JVM to run with these JVM options.

Reply:
> Just to double check, will this logic need to be re-used anywhere? I don't think the executor pods have to mount the same properties config map.

Currently, I don't think so, since executors aren't started with spark-submit. That would be nice to do at some point, to simplify how executors are started, but that's a little beyond what this PR is doing.